src/interpreter.rs

```diff
@@ -1,148 +1,128 @@
+
+        Ok(Self {
+            instructions,
+            instruction_pointer: 0,
+            data_pointer: 0,
+            tape: [0; 3000],
+        })
     }
-    
-    pub fn build_instructions (&mut self) { 
-        /* Converts the code string into a vector of instructions  */
-        
-        self.build_loop_map();
-        for (i, c) in self.code.chars().enumerate() {
-            let i: u32 = i as u32;
-            if c == '+' {
-                self.instructions.push(Instruction::Increment);
-            
-            } else if c == '-' {
-                self.instructions.push(Instruction::Decrement);
-            
-            } else if c == '<' {
-                self.instructions.push(Instruction::MoveLeft);
-
-            } else if c == '>' {
-                self.instructions.push(Instruction::MoveRight);
-            
-            } else if c == '[' {
-                let end = self.loops.get(&i).unwrap();
-                self.instructions.push(Instruction::StartLoop(*end));
-
-            } else if c == ']' {
-                let start = self.loops.get(&i).unwrap();
-                self.instructions.push(Instruction::EndLoop(*start));
-            
-            } else if c == '.' {
-                self.instructions.push(Instruction::Output);
-            
-            } else if c == ',' {
-                self.instructions.push(Instruction::Input);
-            }
-        }
-        self.instructions.push(Instruction::Halt);
+
+    // Loops over the vec of instructions and executes the corresponding code
+    pub fn run(&mut self, mut input: impl Iterator<Item = u8>) {
+        while self.step(&mut input) {}
     }
 
-    pub fn run(&mut self) {
-        /* Loops over the vec of instructions and executes the corresponding code */
-        let tape_size: usize = self.tape.len() - 1; // Represents the last index of the tape rather than the len
-        let mut input_pointer: usize = 0;
-        let inputs: Vec<u8> = self.input.clone().into_bytes();
-        let mut instruction_pointer: usize = 0;
-        loop {
-            let instruction = &self.instructions[instruction_pointer]; 
-            let cell = &mut self.tape[self.pointer]; 
-            match *instruction {
-                Instruction::Increment => {
-                    *cell = cell.overflowing_add(1).0; 
-                },
-                Instruction::Decrement => {
-                    *cell = cell.overflowing_sub(1).0;
-                },
-                Instruction::MoveRight => {
-                    self.pointer = self.pointer.overflowing_sub(1).0;
-                    if self.pointer > tape_size {
-                        self.pointer = tape_size;
-                    }
-                }
-                Instruction::MoveLeft => {
-                    self.pointer += 1;
-                    if self.pointer > tape_size {
-                        self.pointer = 0;
-                    }
-                },

```

Comment:
I believe you flipped what MoveRight and MoveLeft are supposed to do, the if in MoveLeft is also confused